lib/ecto/adapters/sql.ex

```diff
@@ -60,7 +65,22 @@ defmodule Ecto.Adapters.SQL do
       end
 
       @doc false
-      def insert(repo, source, params, returning, opts) do
+      # Nil ids are generated in the database.
+      def insert(repo, source, params, {key, :id, nil}, returning, opts) do
```

Comment:
@michalmuskala Note `insert/6` receives a new argument now. You will need to check for :id and :binary_id in your mongo adapter. Let's talk about this whenever you can.